onmt/utils/loss.py

```diff
@@ -248,10 +250,30 @@ def _make_shard_state(self, batch, output, range_, attns=None):
                 "std_attn": attns.get("std"),
                 "coverage_attn": coverage
             })
+        if self.lambda_align != 0.0:
+            # attn_align should be in (batch_size, pad_tgt_size, pad_src_size)
+            attn_align = attns.get("align", None)
+            # align_idx should be a Tensor in size([N, 3]) of [B, tgt+1, src]
+            align_idx = batch.align
+            assert attns is not None
+            assert attn_align is not None, "lambda_align != 0.0 requires " \
+                "alignement attention head"
+            assert align_idx is not None, "lambda_align != 0.0 requires " \
+                "provide guided alignement"
+            pad_tgt_size, batch_size, _ = batch.tgt.size()
+            pad_src_size = batch.src[0].size(0)
+            align_matrix_size = [batch_size, pad_tgt_size, pad_src_size]
+            ref_align = onmt.utils.make_batch_align_matrix(
+                align_idx, align_matrix_size, normalize=True)
+            # NOTE: tgt-src alignement that in range_, not take count in bos
```

Comment:
?

Reply:
By calling make_batch_align_matrix, the sparse align_idx will be converted to a full alignment matrix. To make it coherent with shard_state['target'], we return a submatrix in the same range of shard_state['target'],.